src/sophia/utility.f

```diff
@@ -0,0 +1,69 @@
+      SUBROUTINE TOEVT()
+C************************************************************************
+C
+C     Converts data in S_PLIST common block to
+C     standard HEPEVT common block format, 
+C     and S_PLIST1 and S_CHP to SCHG common block containing
+C     charge and parent decayed particle information
+C     
+C************************************************************************
+      IMPLICIT DOUBLE PRECISION (A-H,O-Z)
+      IMPLICIT INTEGER (I-N)
+
+      SAVE
+      INTEGER PDGCOD(49)
+      DATA PDGCOD /
+     &  22, -11, 11, -13, 13, 111, 211, -211, 321, -321, 130, 310, 2212,
+     &  2112, 12, -12, 14, -14, -99999999, -99999999, 311, -311, 221, 
+     &  331, 213, -213, 113, 323, -323, 313, -313, 223, 333, 3222, 3212,
+     &  3112, 3322, 3312, 3122, 2224, 2214, 2114, 1114, 3224, 3214, 
+     &  3114, 3324, 3314, 3334 /
+       
+      COMMON /S_PLIST/ P(2000,5), LLIST(2000), NP, Ideb
+      COMMON /S_PLIST1/ LLIST1(2000)
+      COMMON /S_CHP/ S_LIFE(49), ICHP(49), ISTR(49), IBAR(49)
+        
+C     Standard HEPEVT common block:
+      PARAMETER (NMXHEP = 2000)
+      PARAMETER (NMXHP2 = NMXHEP * 2, NMXHP4 = NMXHEP * 4, 
+     & NMXHP5 = NMXHEP * 5)
+      COMMON /HEPEVT/ NEVHEP, NHEP, ISTHEP(NMXHEP), IDHEP(NMXHEP),
+     & JMOHEP(2, NMXHEP), JDAHEP(2, NMXHEP), 
+     & PHEP(5, NMXHEP), VHEP(4, NMXHEP)
+
+C     Initialization of HEVEVT variables
+      DATA NEVHEP / 0 /
+      DATA JMOHEP/ NMXHP2 * 0.0/, JDAHEP / NMXHP2 * 0.0 /
+      DATA PHEP/ NMXHP5 * 0.0/, VHEP / NMXHP4 * 0.0 /
+
+C     Common block SCHG with additional information:
+C     ICHG - charge, 
+C     DECPAR - position in the arrays (ISTHEP, PHEP, ...) of parent (decayed) particle     
+      INTEGER ICHG, DECPAR
+      COMMON /SCHG/ ICHG(NMXHEP), DECPAR(NMXHEP)
+
+      INTEGER CID  ! id of the current (in the loop) particle
+C     number added to decayed particle:
+      INTEGER DNUM
+      PARAMETER (DNUM = 10000)
+       
+      NEVHEP = NEVHEP + 1
+      NHEP = NP
+C     We do not transpose P, because we don't use PHEP         
+C      PHEP = TRANSPOSE(P)
+        
+      DO I = 1, NHEP
+        CID = LLIST(I)
+        IF (ABS(CID) .GT. DNUM) THEN
+            ISTHEP(I) = 2
+            CID = CID - ISIGN(DNUM, CID)
+        ELSE 
+            ISTHEP(I) = 1
+        END IF     
+        IDHEP(I) = ISIGN(1, CID) * PDGCOD(ABS(CID))
```

Comment:
What does it do for neutral particles or K0 like 310, 130?

Reply:
For neutral particles sophia ID is positive, and we have effectively IDHEP(I) = PDGCOD(CID)